cmd/cert-checker/main.go

```diff
@@ -0,0 +1,217 @@
+	} else {
+		// Check we have the right expiration time
+		if parsedCert.NotAfter != cert.Expires {
+			problems = append(problems, "Stored expiration doesn't match certificate NotAfter")
+		}
+		// Check basic constraints are set
+		if !parsedCert.BasicConstraintsValid {
+			problems = append(problems, "Certificate doesn't have basic constraints set")
+		}
+		// Check the cert isn't able to sign other certificates
+		if parsedCert.IsCA {
+			problems = append(problems, "Certificate can sign other certificates")
+		}
+		// Check the cert has the correct validity period
+		if parsedCert.NotAfter.Sub(cert.Issued) > (time.Hour * 24 * 90) {
+			problems = append(problems, "Certificate has a validity period longer than 90 days")
+		}
+		// Check that the PA is still willing to issue for each name in DNSNames + CommonName
+		for _, name := range append(parsedCert.DNSNames, parsedCert.Subject.CommonName) {
+			if err = c.pa.WillingToIssue(core.AcmeIdentifier{Type: core.IdentifierDNS, Value: name}); err != nil {
+				problems = append(problems, fmt.Sprintf("Policy Authority isn't willing to issue for %s: %s", name, err))
+			}
+		}
+		// Check the cert has the correct key usage extensions
+		if !core.CmpExtKeyUsageSlice(parsedCert.ExtKeyUsage, []x509.ExtKeyUsage{x509.ExtKeyUsageServerAuth, x509.ExtKeyUsageClientAuth}) {
+			problems = append(problems, "Certificate has incorrect key usage extensions")
+		}
+	}
+	return problems
+}
+
+func main() {
+	app := cmd.NewAppShell("cert-checker")
+	app.App.Flags = append(app.App.Flags, cli.IntFlag{
+		Name:  "workers",
+		Value: 5,
+		Usage: "The number of cocurrent workers used to process certificates",
+	}, cli.StringFlag{
+		Name:  "report-path",
+		Usage: "The path to write a JSON report on the certificates checks to (if no path is provided the report will not be written out)",
+	}, cli.StringFlag{
+		Name:  "sql-uri",
+		Usage: "SQL URI if not provided in the configuration file",
+	})
+
+	app.Config = func(c *cli.Context, config cmd.Config) cmd.Config {
+		config.CertChecker.ReportDirectoryPath = c.GlobalString("report-dir-path")
+
+		if connect := c.GlobalString("sql-uri"); connect != "" {
+			config.CertChecker.DBConnect = connect
+		}
+
+		return config
+	}
+
+	app.Action = func(c cmd.Config) {
+		stats, err := statsd.NewClient(c.Statsd.Server, c.Statsd.Prefix)
+		cmd.FailOnError(err, "Couldn't connect to statsd")
+
+		auditlogger, err := blog.Dial(c.Syslog.Network, c.Syslog.Server, c.Syslog.Tag, stats)
+		cmd.FailOnError(err, "Could not connect to Syslog")
+
+		blog.SetAuditLogger(auditlogger)
+		auditlogger.Info(app.VersionString())
+
+		dbMap, err := sa.NewDbMap(c.CertChecker.DBConnect)
+		cmd.FailOnError(err, "Could not connect to database")
+
+		checker := newChecker(dbMap)
+		auditlogger.Info("# Getting certificates issued in the last 90 days")
+		err = checker.getCerts()
+		cmd.FailOnError(err, "Failed to get sample certificates")
+
+		if c.CertChecker.Workers > len(checker.certs) {
```

Comment:
Why is this necessary? If there are too many workers, some should just exit when the reader closes the channel, right?